src/hotspot/cpu/x86/peephole_x86.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#include "precompiled.hpp"
+
+#ifdef COMPILER2
+
+#include "opto/peephole.hpp"
+
+#ifdef _LP64
+// This function transforms the shapes
+// mov d, s1; add d, s2 into
+// lea d, [s1 + s2]     and
+// mov d, s1; shl d, s2 into
+// lea d, [s1 << s2]    with s2 = 1, 2, 3
+bool lea_coalesce_helper(Block* block, int block_index, PhaseCFG* cfg_, PhaseRegAlloc* ra_,
+                         MachNode* (*new_root)(), uint inst0_rule, bool imm) {
+  MachNode* inst0 = block->get_node(block_index)->as_Mach();
+  assert(inst0->rule() == inst0_rule, "sanity");
+
+  OptoReg::Name dst = ra_->get_reg_first(inst0);
+  MachNode* inst1 = nullptr;
+  OptoReg::Name src1 = OptoReg::Bad;
+
+  if (inst0->in(1)->is_MachSpillCopy()) {
+    OptoReg::Name in = ra_->get_reg_first(inst0->in(1)->in(1));
+    if (OptoReg::is_reg(in) && OptoReg::as_VMReg(in)->is_Register()) {
+      inst1 = inst0->in(1)->as_Mach();
+      src1 = in;
+    }
```

Comment:
I don't understand why this optimization requires MachSpillCopy.  Is that the only time we sould see mov+add or mov+shift?

Reply:
OK.